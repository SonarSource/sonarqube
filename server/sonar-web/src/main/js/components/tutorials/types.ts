/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export enum TutorialModes {
  Local = 'local',
  Jenkins = 'jenkins',
  BitbucketPipelines = 'bitbucket-pipelines',
  GitLabCI = 'gitlab-ci',
  GitHubActions = 'github-actions',
  AzurePipelines = 'azure-pipelines',
  OtherCI = 'other-ci',
}

export enum BuildTools {
  Maven = 'maven',
  Gradle = 'gradle',
  Cpp = 'cpp',
  ObjectiveC = 'objectivec',
  DotNet = 'dotnet',
  Dart = 'dart',
  Other = 'other',
}

export enum GradleBuildDSL {
  Groovy = 'build.gradle',
  Kotlin = 'build.gradle.kts',
}

export enum OSs {
  Linux = 'linux',
  Windows = 'win',
  MacOS = 'mac',
}

export enum Arch {
  X86_64 = 'x86_64',
  Arm64 = 'arm64',
}

export enum AutoConfig {
  Automatic = 'automatic',
  Manual = 'manual',
}

export type TutorialConfig = {
  autoConfig?: AutoConfig;
  buildTool?: BuildTools;
};
