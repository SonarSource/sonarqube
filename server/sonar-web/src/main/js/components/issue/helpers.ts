/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { IssueTransition } from '../../types/issues';

export function isTransitionDeprecated(transition: IssueTransition) {
  return transition === IssueTransition.Confirm || transition === IssueTransition.Resolve;
}

export function isTransitionHidden(transition: IssueTransition) {
  return transition === IssueTransition.WontFix;
}

export function transitionRequiresComment(transition: IssueTransition) {
  return [IssueTransition.Accept, IssueTransition.FalsePositive].includes(transition);
}
