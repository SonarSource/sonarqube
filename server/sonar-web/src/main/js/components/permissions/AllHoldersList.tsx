/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import * as React from 'react';
import { BasicSeparator, Spinner } from '~design-system';
import {
  Paging,
  PermissionDefinition,
  PermissionDefinitionGroup,
  PermissionGroup,
  PermissionUser,
} from '../../types/types';
import ListFooter from '../controls/ListFooter';
import HoldersList from './HoldersList';
import SearchForm, { FilterOption } from './SearchForm';

interface Props {
  filter: FilterOption;
  groups: PermissionGroup[];
  groupsPaging?: Paging;
  isProjectManaged?: boolean;
  loading?: boolean;
  onFilter: (filter: string) => void;
  onGrantPermissionToGroup: (group: string, permission: string) => Promise<void>;
  onGrantPermissionToUser: (user: string, permission: string) => Promise<void>;
  onLoadMore: () => void;
  onQuery: (query: string) => void;
  onRevokePermissionFromGroup: (group: string, permission: string) => Promise<void>;
  onRevokePermissionFromUser: (user: string, permission: string) => Promise<void>;
  onSelectPermission?: (permissions?: string) => void;
  permissions: Array<PermissionDefinition | PermissionDefinitionGroup>;
  query: string;
  selectedPermission?: string;
  users: PermissionUser[];
  usersPaging?: Paging;
}

export default class AllHoldersList extends React.PureComponent<Props> {
  handleToggleUser = (user: PermissionUser, permission: string) => {
    const hasPermission = user.permissions.includes(permission);

    if (hasPermission) {
      return this.props.onRevokePermissionFromUser(user.login, permission);
    }
    return this.props.onGrantPermissionToUser(user.login, permission);
  };

  handleToggleGroup = (group: PermissionGroup, permission: string) => {
    const hasPermission = group.permissions.includes(permission);

    if (hasPermission) {
      return this.props.onRevokePermissionFromGroup(group.name, permission);
    }

    return this.props.onGrantPermissionToGroup(group.name, permission);
  };

  getPaging = () => {
    const { filter, groups, groupsPaging, users, usersPaging } = this.props;

    let count = 0;
    let total = 0;
    if (filter !== 'users') {
      count += groups.length;
      total += groupsPaging ? groupsPaging.total : groups.length;
    }
    if (filter !== 'groups') {
      count += users.length;
      total += usersPaging ? usersPaging.total : users.length;
    }

    return { count, total };
  };

  render() {
    const {
      filter,
      query,
      groups,
      users,
      permissions,
      selectedPermission,
      loading = false,
      isProjectManaged,
    } = this.props;
    const { count, total } = this.getPaging();

    return (
      <>
        <div>
          <div className="sw-flex sw-justify-between">
            <SearchForm
              filter={filter}
              onFilter={this.props.onFilter}
              onSearch={this.props.onQuery}
              query={query}
            />
            <Spinner loading={loading} />
          </div>
          <BasicSeparator className="sw-mt-4" />
        </div>
        <HoldersList
          isProjectManaged={!!isProjectManaged}
          loading={loading}
          filter={filter}
          groups={groups}
          onSelectPermission={this.props.onSelectPermission}
          onToggleGroup={this.handleToggleGroup}
          onToggleUser={this.handleToggleUser}
          permissions={permissions}
          query={query}
          selectedPermission={selectedPermission}
          users={users}
        />
        <ListFooter count={count} loadMore={this.props.onLoadMore} total={total} />
      </>
    );
  }
}
