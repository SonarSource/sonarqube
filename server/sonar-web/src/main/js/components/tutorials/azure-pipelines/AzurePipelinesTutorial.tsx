/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import * as React from 'react';
import { BasicSeparator, Title, TutorialStep, TutorialStepList } from '~design-system';
import { translate } from '../../../helpers/l10n';
import { AlmKeys } from '../../../types/alm-settings';
import { Component } from '../../../types/types';
import { LoggedInUser } from '../../../types/users';
import AllSet from '../components/AllSet';
import { TutorialConfig } from '../types';
import BranchAnalysisStepContent from './BranchAnalysisStepContent';
import ExtensionInstallationStepContent from './ExtensionInstallationStepContent';
import ServiceEndpointStepContent from './ServiceEndpointStepContent';

export interface AzurePipelinesTutorialProps {
  alm?: AlmKeys;
  baseUrl: string;
  component: Component;
  currentUser: LoggedInUser;
  willRefreshAutomatically?: boolean;
}

export enum Steps {
  ExtensionInstallation = 'ExtensionInstallation',
  ServiceEndpoint = 'ServiceEndpoint',
  BranchAnalysis = 'BranchAnalysis',
}

export default function AzurePipelinesTutorial(props: AzurePipelinesTutorialProps) {
  const { alm, baseUrl, component, currentUser, willRefreshAutomatically } = props;

  const [config, setConfig] = React.useState<TutorialConfig>({});
  const [done, setDone] = React.useState<boolean>(false);

  React.useEffect(() => {
    setDone(Boolean(config.buildTool));
  }, [config.buildTool]);

  return (
    <>
      <Title>{translate('onboarding.tutorial.with.azure_pipelines.title')}</Title>

      <TutorialStepList className="sw-mb-10">
        <TutorialStep
          title={translate(
            `onboarding.tutorial.with.azure_pipelines.${Steps.ExtensionInstallation}.title`,
          )}
        >
          <ExtensionInstallationStepContent />
        </TutorialStep>

        <TutorialStep
          title={translate(
            `onboarding.tutorial.with.azure_pipelines.${Steps.ServiceEndpoint}.title`,
          )}
        >
          <ServiceEndpointStepContent
            baseUrl={baseUrl}
            component={component}
            currentUser={currentUser}
          />
        </TutorialStep>

        <TutorialStep
          title={translate(
            `onboarding.tutorial.with.azure_pipelines.${Steps.BranchAnalysis}.title`,
          )}
        >
          <BranchAnalysisStepContent config={config} setConfig={setConfig} component={component} />
        </TutorialStep>

        {done && (
          <>
            <BasicSeparator className="sw-my-10" />
            <AllSet
              alm={alm ?? AlmKeys.Azure}
              willRefreshAutomatically={willRefreshAutomatically}
            />
          </>
        )}
      </TutorialStepList>
    </>
  );
}
