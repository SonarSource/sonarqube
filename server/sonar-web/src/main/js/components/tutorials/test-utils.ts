/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { screen } from '@testing-library/react';
import { byLabelText, byRole, byText } from '~sonar-aligned/helpers/testSelector';
import { BuildTools, GradleBuildDSL, OSs, TutorialModes } from './types';

const CI_TRANSLATE_MAP: Partial<Record<TutorialModes, string>> = {
  [TutorialModes.BitbucketPipelines]: 'bitbucket_pipelines',
  [TutorialModes.GitHubActions]: 'github_action',
  [TutorialModes.GitLabCI]: 'gitlab_ci',
};

interface GetCopyToClipboardValueArgs {
  i?: number;
  inlineSnippet?: boolean;
  name?: string;
}

export function getCopyToClipboardValue({
  i = 0,
  inlineSnippet = false,
  name = 'copy_to_clipboard',
}: GetCopyToClipboardValueArgs = {}) {
  const button = screen.getAllByRole('button', { name })[i];

  return inlineSnippet
    ? button.previousSibling?.firstChild?.textContent
    : button.nextSibling?.firstChild?.textContent;
}

export function getCopyToClipboardHostURLValue({
  i = 0,
  name = 'copy_to_clipboard',
}: Omit<GetCopyToClipboardValueArgs, 'inlineSnippet'> = {}) {
  return screen.getAllByRole('button', { name })[i].nextSibling?.nextSibling?.firstChild
    ?.textContent;
}

export function getCommonNodes(ci: TutorialModes) {
  return {
    secretsStepTitle: byRole('heading', {
      name: `onboarding.tutorial.with.${CI_TRANSLATE_MAP[ci]}.${
        ci === TutorialModes.GitHubActions ? 'create_secret' : 'variables'
      }.title`,
    }),
    ymlFileStepTitle: byRole('heading', {
      name: `onboarding.tutorial.with.${CI_TRANSLATE_MAP[ci]}.yaml.title`,
    }),
    genTokenDialogButton: byRole('button', {
      name: 'onboarding.token.generate.long',
    }),
    tokenNameInput: byRole('textbox', { name: 'onboarding.token.name.label' }),
    expiresInSelect: byRole('combobox', { name: '' }),
    tokenValue: byText('generatedtoken2'),
    linkToRepo: byRole('link', {
      name: new RegExp(
        `onboarding.tutorial.with.${CI_TRANSLATE_MAP[ci]}.${
          ci === TutorialModes.GitHubActions ? 'secret' : 'variables'
        }.intro.link`,
      ),
    }),
    allSetSentence: byText('onboarding.tutorial.ci_outro.done'),
  };
}

export function getTutorialActionButtons() {
  return {
    continueButton: byRole('button', { name: 'continue' }),
    generateTokenButton: byRole('button', { name: 'onboarding.token.generate' }),
    deleteTokenButton: byRole('button', { name: 'onboarding.token.delete' }),
    finishTutorialButton: byRole('button', { name: 'tutorials.finish' }),
  };
}

export function getTutorialBuildButtons() {
  return {
    describeBuildTitle: byLabelText('onboarding.build'),
    mavenBuildButton: byRole('radio', { name: `onboarding.build.${BuildTools.Maven}` }),
    gradleBuildButton: byRole('radio', { name: `onboarding.build.${BuildTools.Gradle}` }),
    gradleDSLButton: (name: GradleBuildDSL) => byRole('radio', { name }),
    dotnetBuildButton: byRole('radio', { name: `onboarding.build.${BuildTools.DotNet}` }),
    cppBuildButton: byRole('radio', { name: `onboarding.build.${BuildTools.Cpp}` }),
    objCBuildButton: byRole('radio', { name: `onboarding.build.${BuildTools.ObjectiveC}` }),
    dartBuildButton: byRole('radio', { name: `onboarding.build.${BuildTools.Dart}` }),
    otherBuildButton: byRole('radio', { name: `onboarding.build.${BuildTools.Other}` }),
    windowsDotnetCoreButton: byRole('radio', {
      name: `onboarding.build.${BuildTools.DotNet}.win_core`,
    }),
    windowsDotnetFrameworkButton: byRole('radio', {
      name: `onboarding.build.${BuildTools.DotNet}.win_msbuild`,
    }),
    linuxDotnetCoreButton: byRole('radio', {
      name: `onboarding.build.${BuildTools.DotNet}.linux_core`,
    }),
    dotnetCoreButton: byRole('radio', {
      name: `onboarding.build.${BuildTools.DotNet}.variant.dotnet_core`,
    }),
    dotnetFrameworkButton: byRole('radio', {
      name: `onboarding.build.${BuildTools.DotNet}.variant.dotnet_framework`,
    }),
    linuxButton: byRole('radio', { name: `onboarding.build.other.os.${OSs.Linux}` }),
    windowsButton: byRole('radio', { name: `onboarding.build.other.os.${OSs.Windows}` }),
    macosButton: byRole('radio', { name: `onboarding.build.other.os.${OSs.MacOS}` }),
    arm64Button: byRole('radio', { name: `onboarding.build.other.architecture.arm64` }),
    x86_64Button: byRole('radio', { name: `onboarding.build.other.architecture.x86_64` }),
    autoConfigAutomatic: byRole('radio', { name: 'onboarding.build.cpp.autoconfig.automatic' }),
    autoConfigManual: byRole('radio', { name: 'onboarding.build.cpp.autoconfig.manual' }),
  };
}
