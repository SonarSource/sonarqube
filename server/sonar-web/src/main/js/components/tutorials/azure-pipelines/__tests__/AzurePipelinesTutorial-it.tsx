/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UserEvent } from '@testing-library/user-event/dist/types/setup/setup';
import UserTokensMock from '../../../../api/mocks/UserTokensMock';
import { mockComponent } from '../../../../helpers/mocks/component';
import { mockLanguage, mockLoggedInUser } from '../../../../helpers/testMocks';
import { RenderContext, renderApp } from '../../../../helpers/testReactTestingUtils';
import { Permissions } from '../../../../types/permissions';
import { TokenType } from '../../../../types/token';
import { getCopyToClipboardValue, getTutorialBuildButtons } from '../../test-utils';
import { OSs } from '../../types';
import AzurePipelinesTutorial, { AzurePipelinesTutorialProps } from '../AzurePipelinesTutorial';

jest.mock('../../../../api/settings', () => ({
  getAllValues: jest.fn().mockResolvedValue([]),
}));

let tokenMock: UserTokensMock;

beforeAll(() => {
  tokenMock = new UserTokensMock();
});

afterEach(() => {
  tokenMock.reset();
});

it('should render correctly and allow token generation', async () => {
  renderAzurePipelinesTutorial();
  const user = userEvent.setup();

  expect(
    screen.getByRole('heading', { name: 'onboarding.tutorial.with.azure_pipelines.title' }),
  ).toBeInTheDocument();

  // Default step.
  assertDefaultStepIsCorrectlyRendered();

  // Token step.
  assertServiceEndpointStepIsCorrectlyRendered();

  // Generate a token.
  await clickButton(user, 'onboarding.token.generate.long');
  const modal = screen.getByRole('dialog');
  await clickButton(user, 'onboarding.token.generate', modal);
  const lastToken = tokenMock.getLastToken();

  expect(lastToken).toBeDefined();

  expect(lastToken!.type).toBe(TokenType.Global);
  expect(
    within(modal).getByText(`users.tokens.new_token_created.${lastToken!.token}`),
  ).toBeInTheDocument();
  await clickButton(user, 'continue', modal);

  // Analysis step: .NET
  await user.click(getTutorialBuildButtons().dotnetBuildButton.get());
  assertDotNetStepIsCorrectlyRendered();

  // Analysis step: Maven
  await user.click(getTutorialBuildButtons().mavenBuildButton.get());
  assertMavenStepIsCorrectlyRendered();

  // Analysis step: Gradle
  await user.click(getTutorialBuildButtons().gradleBuildButton.get());
  assertGradleStepIsCorrectlyRendered();

  // Analysis step: C Family
  await user.click(getTutorialBuildButtons().cppBuildButton.get());
  // Default: Automatic configuration
  // expect linux/win/macos buttons not to be present
  expect(getTutorialBuildButtons().linuxButton.query()).not.toBeInTheDocument();
  expect(getTutorialBuildButtons().windowsButton.query()).not.toBeInTheDocument();
  expect(getTutorialBuildButtons().macosButton.query()).not.toBeInTheDocument();
  assertAutomaticCppStepIsCorrectlyRendered();

  // Switch to manual configuration
  await user.click(getTutorialBuildButtons().autoConfigManual.get());
  await user.click(getTutorialBuildButtons().linuxButton.get());
  assertManualCppStepIsCorrectlyRendered(OSs.Linux);
  await user.click(getTutorialBuildButtons().arm64Button.get());
  assertManualCppStepIsCorrectlyRendered(OSs.Linux, 'arm64');
  await user.click(getTutorialBuildButtons().windowsButton.get());
  assertObjCStepIsCorrectlyRendered(OSs.Windows);
  await user.click(getTutorialBuildButtons().macosButton.get());
  assertObjCStepIsCorrectlyRendered(OSs.MacOS);

  // Analysis step: C Family
  await user.click(getTutorialBuildButtons().objCBuildButton.get());
  await user.click(getTutorialBuildButtons().linuxButton.get());
  await user.click(getTutorialBuildButtons().x86_64Button.get());
  assertObjCStepIsCorrectlyRendered(OSs.Linux);
  await user.click(getTutorialBuildButtons().arm64Button.get());
  assertObjCStepIsCorrectlyRendered(OSs.Linux, 'arm64');
  await user.click(getTutorialBuildButtons().windowsButton.get());
  assertObjCStepIsCorrectlyRendered(OSs.Windows);
  await user.click(getTutorialBuildButtons().macosButton.get());
  assertObjCStepIsCorrectlyRendered(OSs.MacOS);

  // Analysis step: Dart
  await user.click(getTutorialBuildButtons().dartBuildButton.get());
  assertOtherStepIsCorrectlyRendered();

  // Analysis step: Other
  await user.click(getTutorialBuildButtons().otherBuildButton.get());
  assertOtherStepIsCorrectlyRendered();

  // Finish tutorial
  assertFinishStepIsCorrectlyRendered();
});

it('should not offer CFamily analysis if the language is not available', () => {
  renderAzurePipelinesTutorial(undefined, { languages: {} });

  expect(getTutorialBuildButtons().dotnetBuildButton.get()).toBeInTheDocument();
  expect(getTutorialBuildButtons().cppBuildButton.query()).not.toBeInTheDocument();
});

function assertDefaultStepIsCorrectlyRendered() {
  expect(
    screen.getByRole('heading', {
      name: 'onboarding.tutorial.with.azure_pipelines.ExtensionInstallation.title',
    }),
  ).toBeInTheDocument();
}

function assertServiceEndpointStepIsCorrectlyRendered() {
  expect(
    screen.getByRole('heading', {
      name: 'onboarding.tutorial.with.azure_pipelines.ServiceEndpoint.title',
    }),
  ).toBeInTheDocument();
  expect(getCopyToClipboardValue({ i: 0, name: 'Copy to clipboard', inlineSnippet: true })).toBe(
    'https://sonarqube.example.com/',
  );
  expect(
    screen.getByRole('button', { name: 'onboarding.token.generate.long' }),
  ).toBeInTheDocument();
}

function assertDotNetStepIsCorrectlyRendered() {
  expect(
    screen.getByRole('heading', {
      name: 'onboarding.tutorial.with.azure_pipelines.BranchAnalysis.title',
    }),
  ).toBeInTheDocument();

  expect(getCopyToClipboardValue({ i: 1, name: 'Copy to clipboard', inlineSnippet: true })).toBe(
    'foo',
  );
}

function assertMavenStepIsCorrectlyRendered() {
  expect(getCopyToClipboardValue({ i: 0, name: 'Copy' })).toMatchSnapshot(
    'maven, copy additional properties',
  );
}

function assertGradleStepIsCorrectlyRendered() {
  expect(getCopyToClipboardValue({ i: 0, name: 'Copy' })).toMatchSnapshot(
    'gradle, copy additional properties',
  );
}

function assertObjCStepIsCorrectlyRendered(os: string, arch: string = 'x86_64') {
  expect(getCopyToClipboardValue({ i: 0, name: 'Copy' })).toMatchSnapshot(
    `objectivec ${os} ${arch}, copy shell script`,
  );
  expect(getCopyToClipboardValue({ i: 1, name: 'Copy to clipboard', inlineSnippet: true })).toBe(
    'foo',
  );
  expect(
    getCopyToClipboardValue({ i: 2, name: 'Copy to clipboard', inlineSnippet: true }),
  ).toMatchSnapshot(`objectivec ${os} ${arch}, copy additional properties`);
  expect(getCopyToClipboardValue({ i: 1, name: 'Copy' })).toMatchSnapshot(
    `objectivec ${os} ${arch}, copy build-wrapper command`,
  );
}

function assertAutomaticCppStepIsCorrectlyRendered() {
  assertOtherStepIsCorrectlyRendered();
}

function assertManualCppStepIsCorrectlyRendered(os: string, arch: string = 'x86_64') {
  expect(getCopyToClipboardValue({ i: 0, name: 'Copy' })).toMatchSnapshot(
    `manual-cpp ${os} ${arch}, copy shell script`,
  );
  expect(getCopyToClipboardValue({ i: 1, name: 'Copy to clipboard', inlineSnippet: true })).toBe(
    'foo',
  );
  expect(
    getCopyToClipboardValue({ i: 2, name: 'Copy to clipboard', inlineSnippet: true }),
  ).toMatchSnapshot(`manual-cpp ${os} ${arch}, copy additional properties`);
  expect(getCopyToClipboardValue({ i: 1, name: 'Copy' })).toMatchSnapshot(
    `manual-cpp ${os} ${arch}, copy build-wrapper command`,
  );
}

function assertOtherStepIsCorrectlyRendered() {
  expect(getCopyToClipboardValue({ i: 1, name: 'Copy to clipboard', inlineSnippet: true })).toBe(
    'foo',
  );
}

function assertFinishStepIsCorrectlyRendered() {
  expect(
    screen.getByRole('heading', {
      name: 'onboarding.tutorial.ci_outro.done',
    }),
  ).toBeInTheDocument();
}

function renderAzurePipelinesTutorial(
  props: Partial<AzurePipelinesTutorialProps> = {},
  { languages = { c: mockLanguage({ key: 'c' }) } }: RenderContext = {},
) {
  return renderApp(
    '/',
    <AzurePipelinesTutorial
      baseUrl="https://sonarqube.example.com/"
      component={mockComponent({ key: 'foo' })}
      currentUser={mockLoggedInUser({ permissions: { global: [Permissions.Scan] } })}
      willRefreshAutomatically
      {...props}
    />,
    { languages },
  );
}

async function clickButton(user: UserEvent, name: string, context?: HTMLElement) {
  if (context) {
    await user.click(within(context).getByRole('button', { name }));
  } else {
    await user.click(screen.getByRole('button', { name }));
  }
}
