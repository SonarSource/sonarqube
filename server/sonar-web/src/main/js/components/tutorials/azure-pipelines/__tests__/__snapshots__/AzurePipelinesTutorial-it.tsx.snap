// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly and allow token generation: gradle, copy additional properties 1`] = `
"# Additional properties that will be passed to the scanner,
# Put one key=value per line, example:
# sonar.exclusions=**/*.bin
sonar.projectKey=foo
sonar.projectName=MyProject
"
`;

exports[`should render correctly and allow token generation: manual-cpp linux arm64, copy additional properties 1`] = `"sonar.cfamily.compile-commands=bw-output/compile_commands.json"`;

exports[`should render correctly and allow token generation: manual-cpp linux arm64, copy build-wrapper command 1`] = `"./build-wrapper-linux-aarch64/build-wrapper-linux-aarch64 --out-dir bw-output <your build command here>"`;

exports[`should render correctly and allow token generation: manual-cpp linux arm64, copy shell script 1`] = `
"curl 'http://localhost/static/cpp/build-wrapper-linux-aarch64.zip' --output build-wrapper.zip
unzip build-wrapper.zip"
`;

exports[`should render correctly and allow token generation: manual-cpp linux x86_64, copy additional properties 1`] = `"sonar.cfamily.compile-commands=bw-output/compile_commands.json"`;

exports[`should render correctly and allow token generation: manual-cpp linux x86_64, copy build-wrapper command 1`] = `"./build-wrapper-linux-x86/build-wrapper-linux-x86-64 --out-dir bw-output <your build command here>"`;

exports[`should render correctly and allow token generation: manual-cpp linux x86_64, copy shell script 1`] = `
"curl 'http://localhost/static/cpp/build-wrapper-linux-x86.zip' --output build-wrapper.zip
unzip build-wrapper.zip"
`;

exports[`should render correctly and allow token generation: maven, copy additional properties 1`] = `
"# Additional properties that will be passed to the scanner,
# Put one key=value per line, example:
# sonar.exclusions=**/*.bin
sonar.projectKey=foo
sonar.projectName=MyProject
"
`;

exports[`should render correctly and allow token generation: objectivec linux arm64, copy additional properties 1`] = `"sonar.cfamily.compile-commands=bw-output/compile_commands.json"`;

exports[`should render correctly and allow token generation: objectivec linux arm64, copy build-wrapper command 1`] = `"./build-wrapper-linux-aarch64/build-wrapper-linux-aarch64 --out-dir bw-output <your build command here>"`;

exports[`should render correctly and allow token generation: objectivec linux arm64, copy shell script 1`] = `
"curl 'http://localhost/static/cpp/build-wrapper-linux-aarch64.zip' --output build-wrapper.zip
unzip build-wrapper.zip"
`;

exports[`should render correctly and allow token generation: objectivec linux x86_64, copy additional properties 1`] = `"sonar.cfamily.compile-commands=bw-output/compile_commands.json"`;

exports[`should render correctly and allow token generation: objectivec linux x86_64, copy build-wrapper command 1`] = `"./build-wrapper-linux-x86/build-wrapper-linux-x86-64 --out-dir bw-output <your build command here>"`;

exports[`should render correctly and allow token generation: objectivec linux x86_64, copy shell script 1`] = `
"curl 'http://localhost/static/cpp/build-wrapper-linux-x86.zip' --output build-wrapper.zip
unzip build-wrapper.zip"
`;

exports[`should render correctly and allow token generation: objectivec mac x86_64, copy additional properties 1`] = `"sonar.cfamily.compile-commands=bw-output/compile_commands.json"`;

exports[`should render correctly and allow token generation: objectivec mac x86_64, copy additional properties 2`] = `"sonar.cfamily.compile-commands=bw-output/compile_commands.json"`;

exports[`should render correctly and allow token generation: objectivec mac x86_64, copy build-wrapper command 1`] = `"./build-wrapper-macosx-x86/build-wrapper-macosx-x86 --out-dir bw-output <your build command here>"`;

exports[`should render correctly and allow token generation: objectivec mac x86_64, copy build-wrapper command 2`] = `"./build-wrapper-macosx-x86/build-wrapper-macosx-x86 --out-dir bw-output <your build command here>"`;

exports[`should render correctly and allow token generation: objectivec mac x86_64, copy shell script 1`] = `
"curl 'http://localhost/static/cpp/build-wrapper-macosx-x86.zip' --output build-wrapper.zip
unzip build-wrapper.zip"
`;

exports[`should render correctly and allow token generation: objectivec mac x86_64, copy shell script 2`] = `
"curl 'http://localhost/static/cpp/build-wrapper-macosx-x86.zip' --output build-wrapper.zip
unzip build-wrapper.zip"
`;

exports[`should render correctly and allow token generation: objectivec win x86_64, copy additional properties 1`] = `"sonar.cfamily.compile-commands=bw-output/compile_commands.json"`;

exports[`should render correctly and allow token generation: objectivec win x86_64, copy additional properties 2`] = `"sonar.cfamily.compile-commands=bw-output/compile_commands.json"`;

exports[`should render correctly and allow token generation: objectivec win x86_64, copy build-wrapper command 1`] = `"build-wrapper-win-x86/build-wrapper-win-x86-64.exe --out-dir bw-output <your build command here>"`;

exports[`should render correctly and allow token generation: objectivec win x86_64, copy build-wrapper command 2`] = `"build-wrapper-win-x86/build-wrapper-win-x86-64.exe --out-dir bw-output <your build command here>"`;

exports[`should render correctly and allow token generation: objectivec win x86_64, copy shell script 1`] = `
"Invoke-WebRequest -Uri 'http://localhost/static/cpp/build-wrapper-win-x86.zip' -OutFile 'build-wrapper.zip'
Expand-Archive -Path 'build-wrapper.zip' -DestinationPath '.'"
`;

exports[`should render correctly and allow token generation: objectivec win x86_64, copy shell script 2`] = `
"Invoke-WebRequest -Uri 'http://localhost/static/cpp/build-wrapper-win-x86.zip' -OutFile 'build-wrapper.zip'
Expand-Archive -Path 'build-wrapper.zip' -DestinationPath '.'"
`;
